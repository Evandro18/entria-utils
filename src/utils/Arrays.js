// @flow
export const isEmpty = (value: Array): boolean => value === undefined || value === null || value.length === 0;
