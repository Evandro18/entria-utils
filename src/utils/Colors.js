// @flow
type RGB = {
  r: number,
  g: number,
  b: number,
};
export const hexToRGBObject = (hex: string): RGB => {
  const result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex);
  if (!result) {
    throw new Error('Bad hex');
  }

  return {
    r: parseInt(result[1], 16),
    g: parseInt(result[2], 16),
    b: parseInt(result[3], 16),
  };
};

export const hexToRGB = (hex: string): string => {
  const rgb = hexToRGBObject(hex);
  return `rgba(${rgb.r},${rgb.g},${rgb.b})`;
};

export const hexToRGBA = (hex: string, alpha: number = 1): string => {
  const rgb = hexToRGBObject(hex);
  return `rgba(${rgb.r},${rgb.g},${rgb.b},${alpha})`;
};
