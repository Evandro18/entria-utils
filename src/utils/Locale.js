// @flow
const LOCAL_STORAGE_LOCALE = 'entria-utils-v1-locale';

export const GENERAL = 'GENERAL';
export const BRAZIL = 'BRAZIL';

export const set = (locale: string = GENERAL) => {
  localStorage.setItem(LOCAL_STORAGE_LOCALE, locale);
};

export const get = (): string => localStorage.getItem(LOCAL_STORAGE_LOCALE) || GENERAL;
