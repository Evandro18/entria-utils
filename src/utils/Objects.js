export function concat(objectOne: Object, objectTwo: Object): Object {
  const concatenated = {
    ...objectOne,
  };

  Object.keys(objectTwo).forEach(key => {
    if (typeof concatenated[key] === 'object') {
      concatenated[key] = {
        ...concatenated[key],
        ...objectTwo[key],
      };
    } else {
      concatenated[key] = objectTwo[key];
    }
  });

  return concatenated;
}
