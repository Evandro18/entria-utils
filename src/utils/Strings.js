// @flow
export const isEmpty = (value: string): boolean => value === undefined || value === null || value === '';

export const isAlpha = (value: string): boolean => /^[a-zA-ZÁÉÍÓÚáéíóúÃÕãõçÇ ]+$/.test(value);

export const isEmail = (value: string): boolean => /^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i.test(value);

export const uppercaseFirstLetter = (text: string): string => `${text.charAt(0).toUpperCase()}${text.slice(1)}`;

export const capitalize = (text: string): string =>
  text.toLowerCase().split(' ').map(word => uppercaseFirstLetter(word)).join(' ');

export const numbers = (value: string): string => value.replace(/\D/g, '');
