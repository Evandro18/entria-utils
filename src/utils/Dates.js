// @flow
import fnsIsValid from 'date-fns/is_valid';
import fnsIsFuture from 'date-fns/is_future';
import fnsIsPast from 'date-fns/is_past';
import fnsCompareAsc from 'date-fns/compare_asc';

import * as Locale from './Locale';

type PossibleDate = Date | string;

type GeneralConfig = {
  locale?: string,
};
const sanitizeConfig = (config: GeneralConfig): GeneralConfig => {
  const sanitizedConfig = { ...config };

  if (!sanitizedConfig.locale) {
    sanitizedConfig.locale = Locale.get();
  }

  return sanitizedConfig;
};

type ExtractedDate = {
  day: string,
  month: string,
  year: string,
};
export function extract(
  value: string,
  config: GeneralConfig = {}
): ExtractedDate {
  const sanitizedConfig = sanitizeConfig(config);

  // brazil = dd/mm/yyyy
  if (sanitizedConfig.locale === Locale.BRAZIL) {
    return {
      day: value.substr(0, 2),
      month: value.substr(3, 2),
      year: value.substr(6, 4),
    };
  }

  // general = yyyy-mm-dd
  return {
    day: value.substr(8, 2),
    month: value.substr(5, 2),
    year: value.substr(0, 4),
  };
}

export function parse(
  value: PossibleDate,
  config: GeneralConfig = {}
): Date {
  if (value && typeof value === 'string') {
    const { day, month, year } = extract(value, config);
    return new Date(year, parseInt(month, 10) - 1, day);
  }

  return new Date(value);
}

export function isValid(
  date: PossibleDate,
  config: GeneralConfig = {}
): boolean {
  const isValidFns = fnsIsValid(parse(date, config));

  const { day, month, year } = extract(date, config);
  const toCompare = new Date(year, parseInt(month - 1, 10), day, 0, 0, 0, 0);
  const dayMatches = toCompare.getDate() === parseInt(day, 10);
  const monthMatches = toCompare.getMonth() + 1 === parseInt(month, 10);
  const yearMatches = toCompare.getFullYear() === parseInt(year, 10);

  return isValidFns && dayMatches && monthMatches && yearMatches;
}

export function isFuture(
  date: PossibleDate,
  config: GeneralConfig = {}
): boolean {
  return fnsIsFuture(parse(date, config));
}

export function isPast(
  date: PossibleDate,
  config: GeneralConfig = {}
): boolean {
  return fnsIsPast(parse(date, config));
}

export function compare(
  dateLeft: PossibleDate,
  dateRight: PossibleDate,
  config: GeneralConfig = {}
): number {
  return fnsCompareAsc(parse(dateLeft, config), parse(dateRight, config));
}

export function isEmpty(
  date: PossibleDate,
  config: GeneralConfig = {}
): boolean {
  return date === undefined || date === null || !fnsIsValid(parse(date, config));
}
