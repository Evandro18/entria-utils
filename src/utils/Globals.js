// @flow
import * as Arrays from './Arrays';
import * as Objects from './Objects';
import * as Strings from './Strings';

export const isEmpty = (value: any) => {
  if (typeof value === 'undefined' || value === null) {
    return true;
  }

  if (Array.isArray(value)) {
    return Arrays.isEmpty(value);
  }

  if (typeof value === 'object') {
    return Objects.isEmpty(value);
  }

  if (typeof value === 'string') {
    return Strings.isEmpty(value);
  }

  return false;
};
