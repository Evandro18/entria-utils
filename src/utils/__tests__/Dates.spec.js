import { setupTest } from '../../../test/utils';
import * as Dates from '../Dates';
import * as Locale from '../Locale';

beforeEach(async () => setupTest());

const brazilConfig = {
  locale: Locale.BRAZIL,
};

it('extract', () => {
  expect(Dates.extract('2000-01-01')).toMatchSnapshot();
});

it('extract - Locale.BRAZIL', () => {
  expect(Dates.extract('01/01/2000', brazilConfig)).toMatchSnapshot();
});

it('parse', () => {
  expect(Dates.parse('2000-01-01')).toMatchSnapshot();
});

it('parse - Locale.BRAZIL', () => {
  expect(Dates.parse('01/01/2000', brazilConfig)).toMatchSnapshot();
});

it('parse - another date', () => {
  expect(Dates.parse(new Date('01/01/2000'))).toMatchSnapshot();
});

it('isValid - valid date', () => {
  expect(Dates.isValid('2000-01-01')).toMatchSnapshot();
});

it('isValid - invalid date', () => {
  expect(Dates.isValid('2000-01-35')).toMatchSnapshot();
});

it('isValid - valid leap date', () => {
  expect(Dates.isValid('2000-02-29')).toMatchSnapshot();
});

it('isValid - invalid leap date', () => {
  expect(Dates.isValid('2001-02-29')).toMatchSnapshot();
});

it('isValid - Locale.BRAZIL - valid date', () => {
  expect(Dates.isValid('01/01/2000', brazilConfig)).toMatchSnapshot();
});

it('isValid - Locale.BRAZIL - invalid date', () => {
  expect(Dates.isValid('35/01/2000', brazilConfig)).toMatchSnapshot();
});

it('isValid - Locale.BRAZIL - valid leap date', () => {
  expect(Dates.isValid('29/02/2000', brazilConfig)).toMatchSnapshot();
});

it('isValid - Locale.BRAZIL - invalid leap date', () => {
  expect(Dates.isValid('29/02/2001', brazilConfig)).toMatchSnapshot();
});

it('isFuture - future date', () => {
  expect(Dates.isFuture('2050-01-01')).toMatchSnapshot();
});

it('isFuture - past date', () => {
  expect(Dates.isFuture('2000-01-01')).toMatchSnapshot();
});

it('isFuture - Locale.BRAZIL - future date', () => {
  expect(Dates.isFuture('01/01/2050', brazilConfig)).toMatchSnapshot();
});

it('isFuture - Locale.BRAZIL - past date', () => {
  expect(Dates.isFuture('01/01/2000', brazilConfig)).toMatchSnapshot();
});

it('isPast - past date', () => {
  expect(Dates.isPast('2000-01-01')).toMatchSnapshot();
});

it('isPast - future date', () => {
  expect(Dates.isPast('2050-01-01')).toMatchSnapshot();
});

it('isPast - Locale.BRAZIL - past date', () => {
  expect(Dates.isPast('01/01/2000', brazilConfig)).toMatchSnapshot();
});

it('isPast - Locale.BRAZIL - future date', () => {
  expect(Dates.isPast('01/01/2050', brazilConfig)).toMatchSnapshot();
});

it('compare - left date greater', () => {
  expect(Dates.compare('2010-01-01', '2000-01-01')).toMatchSnapshot();
});

it('compare - right date greater', () => {
  expect(Dates.compare('2000-01-01', '2010-01-01')).toMatchSnapshot();
});

it('compare - same dates', () => {
  expect(Dates.compare('2000-01-01', '2000-01-01')).toMatchSnapshot();
});

it('compare - Locale.BRAZIL - left date greater', () => {
  expect(Dates.compare('01/01/2010', '01/01/2000', brazilConfig)).toMatchSnapshot();
});

it('compare - Locale.BRAZIL - right date greater', () => {
  expect(Dates.compare('01/01/2000', '01/01/2010', brazilConfig)).toMatchSnapshot();
});

it('compare - Locale.BRAZIL - same dates', () => {
  expect(Dates.compare('01/01/2000', '01/01/2000', brazilConfig)).toMatchSnapshot();
});
