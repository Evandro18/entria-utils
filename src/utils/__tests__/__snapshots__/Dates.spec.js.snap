// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compare - Locale.BRAZIL - left date greater 1`] = `1`;

exports[`compare - Locale.BRAZIL - right date greater 1`] = `-1`;

exports[`compare - Locale.BRAZIL - same dates 1`] = `0`;

exports[`compare - left date greater 1`] = `1`;

exports[`compare - right date greater 1`] = `-1`;

exports[`compare - same dates 1`] = `0`;

exports[`extract - Locale.BRAZIL 1`] = `
Object {
  "day": "01",
  "month": "01",
  "year": "2000",
}
`;

exports[`extract 1`] = `
Object {
  "day": "01",
  "month": "01",
  "year": "2000",
}
`;

exports[`isFuture - Locale.BRAZIL - future date 1`] = `true`;

exports[`isFuture - Locale.BRAZIL - past date 1`] = `false`;

exports[`isFuture - future date 1`] = `true`;

exports[`isFuture - past date 1`] = `false`;

exports[`isPast - Locale.BRAZIL - future date 1`] = `false`;

exports[`isPast - Locale.BRAZIL - past date 1`] = `true`;

exports[`isPast - future date 1`] = `false`;

exports[`isPast - past date 1`] = `true`;

exports[`isValid - Locale.BRAZIL - invalid date 1`] = `false`;

exports[`isValid - Locale.BRAZIL - invalid leap date 1`] = `false`;

exports[`isValid - Locale.BRAZIL - valid date 1`] = `true`;

exports[`isValid - Locale.BRAZIL - valid leap date 1`] = `true`;

exports[`isValid - invalid date 1`] = `false`;

exports[`isValid - invalid leap date 1`] = `false`;

exports[`isValid - valid date 1`] = `true`;

exports[`isValid - valid leap date 1`] = `true`;

exports[`parse - Locale.BRAZIL 1`] = `2000-01-01T02:00:00.000Z`;

exports[`parse - another date 1`] = `2000-01-01T02:00:00.000Z`;

exports[`parse 1`] = `2000-01-01T02:00:00.000Z`;
